### ISSUE TITLE
broken readln in UI listener

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1281]
From the mailing list:

> How to read multiple lines from standard input (in Listener)?
>
> I tried to write something like "readln readln", then copy-paste two lines of text to Listener, but I got an unexpected result. I got first line on stack, then program waited for my input. I pressed the "Enter" key and, of course, got empty string on stack. So, I got "Line1" "" instead of "Line1" "Line2". How to get both lines? For example, "Line1\nLine2" or something else. Maybe I chose the wrong input method?

**Comment by bjourne** [comment]
Works in the console and in FUEL, so the UI must be doing it wrong.

**Comment by mrjbq7** [comment]
Yep, I think it's something to do with ``M\ interactor stream-readln``.

**Comment by erg** [comment]
It's that ``interactor-read`` calls ``interactor-finish`` which calls ``clear-editor``. Instead, clear-editor should be called outside of the read code.

I don't know why ``ctrl-d`` makes it always return ``f`` though.

```factor
: interactor-yield ( interactor -- obj )
    dup thread>> self eq? [
        {
            [ t >>waiting drop ]
            [ flag>> raise-flag ]
            [ mailbox>> mailbox-get ]
            [ f >>waiting drop ]
        } cleave
    ] [ drop f ] if ;

: interactor-finish ( interactor -- )
    [ history>> history-add ] keep
    [ print-input ]
    [ clear-editor drop ]
    [ model>> clear-undo drop ] 2tri ;

: interactor-read ( interactor -- lines )
    [ interactor-yield ] [ interactor-finish ] bi ;
```

**Comment by mrjbq7** [comment]
Somehow clear should happen after all the pending code is executed, not as each ``stream-read`` or ``stream-readln`` is called I would guess?

**Comment by erg** [comment]
Here's a good demo of the problem:

``readln readln``

<img width="265" alt="Screen Shot 2021-12-21 at 11 46 38 AM" src="https://user-images.githubusercontent.com/27430/146975285-ea12a3f6-9a8d-4448-afd0-8930e7399ca9.png">

**Comment by Capital-EX** [comment]
Looking into it it seems like `stream-readln` will read all lines at once. I can confirm this by modifying `M: interactor stream-readln` to the following:
```
M: interactor stream-readln
    interactor-read dup . ?first ;
```
When I run a single `readln`, all lines I C-c/C-v will be consumed:
![Screenshot_20230401_001656](https://user-images.githubusercontent.com/13488718/229265247-b3ca23a4-637d-4621-887e-d2519866283e.png)

**Comment by Capital-EX** [comment]
After poking around it seems like the flow of things is 

```
=> evaluate-input get invoked
   => control-value gets current text
      => passed to interactor-continue
         => maibox-put stores value
            => mailbox-get read values
               => interactor resumes 
                  => interactor-finish eventually gets called
                     => input is prematurely cleared
```

I believe the crux of the issues is `interactor-finish` assumes you are being a "good input source" and not leaving the buffer dirty. Therefor, to have parity between the CLI Listener and the GUI Listener, factor has to write back the unused data to the text input field. Which can then be read later or discarded once evaluation is complete. Only when the currently inputted code has finished, should `interactor-finish` be invoke to clear the UI command-line.

tl;dr: `M: interactor stream-readln` and `M: interactor stream-read1` need their own implementations of `interactor-finish`. And, the current `interactor-finish` should only be invoked once execution has been completed.

**Comment by Capital-EX** [comment]
Just putting this here so I don't have to constantly remember it: 

All of the following also behave differently in the CLI vs GUI:

- [ ] readln
- [ ] read1
- [ ] read
- [ ] read-quot
- [ ] read-until

(Tho, I think that might be all of them)
